### PR TITLE
feat: oneshot logs/tail for flox services

### PR DESCRIPTION
**Pull request by ysndr**
### feat: impl ProcessComposeLogTail to read logs with --tail

Empirically, process-compose will
* log the last _up to_ `<n>` lines logged by a process
* exit with `0` exit code
* not wait for more logs of running processes, even if they logged less than `<n>` lines yet
* print a particular error message to `stderr`:
  ```
  $ process-compose process logs -u test.sock -n 2 foo 2>/dev/null
  2
  3
  write close: write unix ->test.sock: write: broken pipe
  ```
  Currently we use this as another cue that the reading completed,
  however we cannot know whether this was printed by process compose or a running service.
  A PR to banish these logs to stderr was opened upstream at
  <https://github.com/F1bonacc1/process-compose/pull/216>

This adds a new `ProcessComposeLogTail` that calls `process-compose process logs` with `--tail`
and directly reads its output, storing all lines in a vector.

Because log lines are currently not timestamped,
we can only meaningfully provide tail logs for a single process.
Thus the complexity of the multiplexed readers we need for streaming logs can be sidestepped here,
with an independent reading implementation and `Command::output`.

### feat: allow configuring tail for log streams and add `--tail` flag

This is a convenience enabled by recent process-compose updates.
By default, `process-compose process logs --follow` will _also_ print the entirety of the currently buffered lines for the logged process.
Since that can be a few, we provide the same `--tail`/`-n` flag we would like for non following logs, to following logs as well.
My memory tells me that previously `process-compose` ignored `-n` when `--follow` was provided.
This also ensures that the interface for following and non-following logs is closer.

### feat: impl one shot log for single processes

Implement the currently stubbed out branch of `!self.follow`.
Ensure that only a single service name was provided and that the service indeed exists.
Without the latter check, we would ask `process-compose` for logs of a non-existent service,
which would block indefinitely without output.

### test: add `logs` integration tests

* Test service name arg handling for oneshot logs and `--follow`
Oneshot logs require exactly one argument while with `--follow`
multiple logs can be specified, or omitted to default to all processes.

* Test whether without `--follow`, `logs` exits by itself
and does not wait for the service to finish or produce more logs.
More detailed tests are found in the respective unit tests for `ProcessComposeLogTail`.
Apparently `process-compose` blocks for another second or so
after printing logs almost instantaneously.
We heuristically account for that with a timeout.

* Test that --follows waits for logs and keeps receiving new lines.
Wrapping these calls in a `timeout` to ensure they are being killed
as they would otherwise run indefinitely.

### CODE REVIEW

File: cli/flox/src/commands/services/logs.rs
Comment (mkenigs) [inline cli/flox/src/commands/services/logs.rs:64]: nit: I think we should probably standardize the wording for this, I added a `service_does_not_exist_error` helper in the `start` PR

File: cli/tests/services.bats
Comment (mkenigs) [inline cli/tests/services.bats:370]:
```suggestion
# Assert that flox is _not_ waiting for the service to finish.
```

File: cli/tests/services.bats
Comment (mkenigs) [inline cli/tests/services.bats:417]: I think that's already checked by the `-124`
```suggestion
```

File: cli/flox-rust-sdk/src/providers/services.rs
Comment (mkenigs) [inline cli/flox-rust-sdk/src/providers/services.rs:932]:
```suggestion
    /// even if the process logs more lines eventually,
```

File: cli/flox-rust-sdk/src/providers/services.rs
Comment (mkenigs) [inline cli/flox-rust-sdk/src/providers/services.rs:480]: nit: is this struct necessary? Seems like we could just have a function that returns `Vec<ProcessComposeLogLine>`

File: cli/tests/services/logging_services.toml
Comment (mkenigs) [inline cli/tests/services/logging_services.toml:13]: nonblocking: there are a lot of very racey tests in both the integration and unit tests. I'm not sure if it's necessary, but it might be good to try to avoid the race. One idea is adding a `touch file` before the `sleep`, and then the tests can wait until that file exists

**Review by mkenigs — APPROVED** [review]

File: cli/tests/services/logging_services.toml
Comment (mkenigs) [reply cli/tests/services/logging_services.toml:13]: And if the test failure was because of a race, we should definitely fix before merging
https://github.com/flox/flox/actions/runs/10302640067/job/28517184738?pr=1903#step:6:391

**Review by dcarley — APPROVED** [review]
LGTM pending test failure.

File: cli/tests/services/logging_services.toml
Comment (ysndr) [reply cli/tests/services/logging_services.toml:17]: fixed sync with @dcarley

File: cli/tests/services.bats
Comment (ysndr) [reply cli/tests/services.bats]: i think assert_success still fails if the exit code was non 0 despite `-124` but i guess might be able to rely on `run` to fail earlier

File: cli/flox/src/commands/services/logs.rs
Comment (ysndr) [reply cli/flox/src/commands/services/logs.rs]: i copied in your changes hope that git figures that out correctly...

**Review by mkenigs — APPROVED** [review]
Changes look good. Looks like there was the same test failure as before. Should we use the same synchronization mechanisms in the unit tests as the integration tests?

**Comment by ysndr** [comment]
> Looks like there was the same test failure as before

yeah, at this point i'm rather considering that this is because we dont get all log lines printed to stdout in less than .25 of a second
to be fair we are stacking 

- our arg parsing,
	- process-composes arg parsing
   		- process-compose connecting to a socket
   			- reading from socket, printing to stderr	
   			
In theory this should all be quick but summing it all up there might be a bit more than .25 of delay in this.

> Should we use the same synchronization mechanisms in the unit tests as the integration tests?

So far the unit tests seem stable with the timeouts we got there.
thanks to the test instance there's also already a "wait for the socket" phase.

There's also some amount of complexity to add to the unit tests to setup and wait for the pipe, i think unless we have good reason to assume they significantly reduce test times, or not having that synchronization causes flakiness, id skew toward simpler tests.

In fact this test failure does not seem to be entirely rooted in not having the synchronization, rather than printing in general taking longer than the timeout allows on that specific platform (which seems to be a slower github runner? its being the same that platform that failed before iirc)

**Comment by ysndr** [comment]
regarding the output i found that apparently defining local variables in process compose doesnt work? I'm looking into that presently

**Review by dcarley — APPROVED** [review]
Updates since last review LGTM.

We can revisit the timeouts if we have any intermittent test failures on main.

**Pull request merged**
